test/queue_test.cpp

```diff
@@ -0,0 +1,102 @@
+// This is a personal academic project. Dear PVS-Studio, please check it.
+// PVS-Studio Static Code Analyzer for C, C++, C#, and Java: http://www.viva64.com
+
+#include <gtest/gtest.h>
+#include <thread>
+#include "queue.hpp"
+
+constexpr size_t NUMBER_OF_THREADS = 16;
+constexpr size_t NUMBER_OF_OPERATIONS = 10e7;
+
+class QueueTest : public ::testing::Test {
+protected:
+    void SetUp() override {
+        q1.enqueue(1);
+        q2.enqueue(2);
+        q2.enqueue(3);
+    }
+    Queue<int> q0;
+    Queue<int> q1;
+    Queue<int> q2;
+};
+
+TEST_F(QueueTest, IsEmptyInitially) {
+    EXPECT_EQ(q0.get_size(), 0);
+}
+
+TEST_F(QueueTest, TryDequeueWorks) {
+    int value;
+    EXPECT_FALSE(q0.try_dequeue(&value));
+
+    ASSERT_TRUE(q1.try_dequeue(&value));
+    EXPECT_EQ(value, 1);
+    EXPECT_EQ(q1.get_size(), 0);
+
+    ASSERT_TRUE(q2.try_dequeue(&value));
+    EXPECT_EQ(value, 2);
+    EXPECT_EQ(q2.get_size(), 1);
+}
+
+void multiple_enqueue(Queue<int>& queue, size_t count) {
+    for (size_t i = 0; i < count; ++i) {
+        queue.enqueue(static_cast<int>(i));
+    }
+}
+
+void multiple_dequeue(Queue<int>& queue, size_t count) {
+    for (size_t i = 0; i < count; ++i) {
+        queue.dequeue();
+    }
+}
+
+void multiple_enqueue_and_dequeue(Queue<int>& queue, size_t count) {
+    for (size_t i = 0; i < count; ++i) {
+        if (i % 5 == 4) {
+            queue.dequeue();
+        } else {
+            queue.enqueue(static_cast<int>(i));
+        }
+    }
+}
+
```

Comment:
I would also propose the test with some sort of randomization, where after each dequeue/enqueue thread sleeps for random amount of time

Reply:
The current approach is over-optimistic in sense of perfomance, as the compiler knows that we will make enqueue/dequeue on each line.